### DISCUSSION TITLE
User Defined Variables

### DISCUSSION THREAD

**Discussion by roninhaf** [SetBased/php-audit#69] [Q&A]
Hi, I'm trying to create a User Defined Variable using the example described here https://php-audit.readthedocs.io/en/stable/miscellaneous.html#explicit-query-from-php

Unfortunately, the result logged is always NULL even though we're able to log everything else with PHP-Audit.

The environment used is CodeIgniter 3.1.11 and PHP 8.1.

I create the query as soon as the user is logged in.

How do you do it?

Any help would be appreciated :)

**Comment by prwater** [comment]
Hello Hafid,

It hard the tell what is the cause of your issue without access to your audit configuration file.

Here are some tips:
* Did you include the user defined variable in de expression for the column like in [example 7](https://php-audit.readthedocs.io/en/stable/audit-config-file.html#example-7-application-session)?
* Rerun the audit command after you have modified the audit configuration file, e.g. `bin/audit audit etc/audit.cfg`.
* Look in the code of the triggers, the user defined variable must be in generated SQL statements.

If you can't solve your issue, share your audit configuration file with me.

Greetings,
Paul

**Reply by roninhaf** [reply]
Hello, thanks for your reply Paul,

Here is the config file [audit.json](https://github.com/SetBased/php-audit/files/13810483/audit.json)

So we use the [example 8](https://php-audit.readthedocs.io/en/stable/audit-config-file.html#example-8-end-user) and yes we rerun the audit command but the configuration file has not been modified recently.

Here is how the user defined variable is triggered in the code:

`protected function user_defined_variable_audit($id)
    {
        $this->db->query(sprintf('SET @audit_usr_id = %s', $id ?? 'null'));
    }`

And the method is called as soon as the user successfully logs in.

**Comment by prwater** [comment]
Everything looks fine. This case is covered by this [unit test](https://github.com/SetBased/php-audit/blob/83c25991ebf1507ec458aa0f469167168b17d281/test/MySql/AuditCommand/PhpPlaisio/PhpPlaisioTest.php#L197). The only difference is `query` vs `real_query`, this should not mater.

* The variable $id is not null?
* Some data has been changed after the user logs in during the same page request?
* At next page requests the variable `@audit_usr_id` must be set again (this time using data stored in the session).

**Reply by roninhaf** [reply]
I can confirm that the variable $id is not null.

When you say at next page request to set the variable again using data stored in the session, are you talking about database session or PHP session?

**Comment by prwater** [comment]
The PHP session. The MySQL user defined variable `@audit_usr_id` will be lost as soon as PHP closes the connection to the MySQL instance (i.e. at the end of serving the page request).
